### DISCUSSION TITLE
Shared subscriptions and clean session false

### DISCUSSION THREAD

**Discussion by teemupiiroinenwirepas** [vernemq/vernemq#1955] [Q&A] [answered]
Hello

I saw the bug (https://github.com/vernemq/vernemq/issues/1728) about shared subscriptions and clean session functionality when clean session is false. If I understand the ticket correctly it is not possible to have variable amount of subscribers to shared subscriptions topic with clean session false without losing some buffered messages. Is this true? Our case is that we would like to have multiple subscribers to shared subscriptions topic, and the amount of those subscribers would vary depending on the load. We could have some workaround by having minimum static amount of subscribers that have clean session false and the others (varying amount of subscribers) have it always true.

Br,
Teemu

**Comment by ioolkos** [comment, marked as answer]
@teemupiiroinenwirepas Hello & welcome.
Thanks for opening this discussion topic.

A couple of quick remarks: 
https://github.com/vernemq/vernemq/issues/1728 is not a bug; but it might be a feature request.

The whole thing might sound confusing but it is quite clear. To simplify: VerneMQ treats Shared subscriber load balancing at the routing level, by choosing (as per SG Policy) one subscriber and delivering the message to that subscriber's queue. Every session has its own queue and the identifier for that session is the ClientID. If we have 5 shared subscribers we have 5 queues.

Note: if a queue belongs to a persistent session (and by that a unique ClientID) it will store messages even when the Client is offline. No other ClientID will ever be able to get to those stored messages.

The feature request in https://github.com/vernemq/vernemq/issues/1728 wants to treat Shared subscriber load balancing as if there is only 1 virtual or actual queue. Getting messages from that queue will be load balanced (queues generally work this way). 
This group-queue could still collect messages when all subscribers are offline, of course. It would also deliver all the stored messages to the first consumer getting back online.

This makes a lot of sense when the subscribers are Kubernetes pods using their pod names as ClientIDs (those are then random). Putting aside the question of respecting the requested QoS and clean_session state, we then treat all those subscribers as "the same". This is intuitive.

Implementing this would require a new queue type. Your next question will be whether this queue is replicated on all cluster nodes :)

Note that the recently added SG Policy `prefer_online_before_local` might be a little bit more aligned with what you might intuitively expect: always prefer delivery to an online Subscriber (even when local offline queues live on the same node where the message comes in). Commit is this one: https://github.com/vernemq/vernemq/commit/e5e87025cacded0684996cb72b9a724d12e41976

Did I manage to explain things a little better?

**Reply by teemupiiroinenwirepas** [reply]
Thanks for the reply. I see that it is not so simple case to support single queue, also thanks for the note about "prefer_online_before_local". That is how I would have thought it would work by default, so need to set that one on.
